src/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEBuilder.cs

```diff
@@ -409,5 +411,162 @@ private static int SumRawDataSizes(ImmutableArray<SerializedSection> sections,Se
+                    }
+                    else if (blob.Buffer == strongNameSignatureFixup.Buffer)
+                    {
+                        yield return GetPrefixBlob(new Blob(blob.Buffer, blobStart, blobLength), strongNameSignatureFixup);
+                        yield return GetSuffixBlob(new Blob(blob.Buffer, blobStart, blobLength), strongNameSignatureFixup);
+                        break;
+                    }
+                    else
+                    {
+                        yield return new Blob(blob.Buffer, blobStart, blobLength);
+                        break;
+                    }
+                }
+            }
+        }
+
+        // internal for testing
+        internal static Blob GetPrefixBlob(Blob container, Blob blob) => new Blob(container.Buffer, container.Start, blob.Start - container.Start);
+        internal static Blob GetSuffixBlob(Blob container, Blob blob) => new Blob(container.Buffer, blob.Start + blob.Length, container.Start + container.Length - blob.Start - blob.Length);
+
+        // internal for testing
+        internal static IEnumerable<Blob> GetContentToChecksum(BlobBuilder peImage, Blob checksumFixup)
+        {
+            foreach (var blob in peImage.GetBlobs())
+            {
+                if (blob.Buffer == checksumFixup.Buffer)
+                {
+                    yield return GetPrefixBlob(blob, checksumFixup);
+                    yield return GetSuffixBlob(blob, checksumFixup);
+                }
+                else
+                {
+                    yield return blob;
+                }
+            }
+        }
+
+        internal void Sign(BlobBuilder peImage, Blob strongNameSignatureFixup, Func<IEnumerable<Blob>, byte[]> signatureProvider)
+        {
+            Debug.Assert(peImage != null);
+            Debug.Assert(signatureProvider != null);
+
+            int peHeadersSize = Header.ComputeSizeOfPEHeaders(GetSections().Length);
+            byte[] signature = signatureProvider(GetContentToSign(peImage, peHeadersSize, Header.FileAlignment, strongNameSignatureFixup));
+
+            // signature may be shorter (the rest of the reserved space is padding):
+            if (signature == null || signature.Length > strongNameSignatureFixup.Length)
+            {
+                throw new InvalidOperationException(SR.SignatureProviderReturnedInvalidSignature);
+            }
+
+            uint checksum = CalculateChecksum(peImage, _lazyChecksum);
+            new BlobWriter(_lazyChecksum).WriteUInt32(checksum);
+
+            var writer = new BlobWriter(strongNameSignatureFixup);
+            writer.WriteBytes(signature);
+        }
+
+        // internal for testing
+        internal static uint CalculateChecksum(BlobBuilder peImage, Blob checksumFixup)
+        {
+            return CalculateChecksum(GetContentToChecksum(peImage, checksumFixup)) + (uint)peImage.Count;
+        }
+
+        private unsafe static uint CalculateChecksum(IEnumerable<Blob> blobs)
```

Comment:
AFAICT, the PE/COFF spec just says "The image file checksum. The algorithm for computing the checksum is incorporated into IMAGHELP.DLL". Should we push to have this algorithm documented? Should we add the documentation to our Addendum to that spec in this repo in the meantime?